armadillo/src/main/java/org/molgenis/armadillo/storage/CharacterSeparatedFile.java

```diff
@@ -0,0 +1,207 @@
+    this.file = file;
+    // test if , is real separator
+    CSVReader reader = this.getReader();
+    this.setHeader(reader.readNext());
+    char separator = getSeparatorFromHeader(header, file);
+    this.setSeparator(separator);
+    // if , is not the separator, change it
+    if (separator != ',') {
+      reader = this.getReader();
+      this.setHeader(reader.readNext());
+    }
+    this.types = this.getTypesFromData(reader);
+    this.setSchema(this.createSchemaFromTypes(this.types, this.header));
+  }
+
+  public void setNumberOfRowsToDetermineTypeBy(int numberOfRowsToDetermineTypeBy) {
+    this.numberOfRowsToDetermineTypeBy = numberOfRowsToDetermineTypeBy;
+  }
+
+  public void setSeparator(char separator) {
+    this.separator = separator;
+  }
+
+  public void setHeader(String[] header) {
+    this.header = header;
+  }
+
+  public void setHeaderColname(String colname, int index) {
+    this.header[index] = colname;
+  }
+
+  static char getSeparatorFromHeader(String[] header, MultipartFile file)
+      throws CsvValidationException {
+    if (header.length == 1) {
+      String headerLine = header[0];
+      if (headerLine.split(";").length > 1) {
+        return ';';
+      } else if (headerLine.split("\t").length > 1) {
+        return '\t';
+      } else if (headerLine.split("\\|").length > 1) {
+        return '|';
+      } else {
+        // separator must be , ; \t or |
+        throw new CsvValidationException(
+            String.format(
+                "Unsupported separator in file [%s] with header: [%s]",
+                file.getOriginalFilename(), headerLine));
+      }
+    } else {
+      return ',';
+    }
+  }
+
+  private CSVReader getReader() throws IOException {
+    CSVParser parser = new CSVParserBuilder().withSeparator(separator).build();
+    return new CSVReaderBuilder(new InputStreamReader(file.getInputStream()))
+        .withCSVParser(parser)
+        .build();
+  }
+
+  static String getTypeOfCell(String cell) {
+    try {
+      Double.parseDouble(cell);
+      return CharacterSeparatedFile.DOUBLE;
+    } catch (NumberFormatException e) {
+      return CharacterSeparatedFile.STRING;
+    }
+  }
+
+  List<String> getTypesFromData(CSVReader reader) throws IOException, CsvValidationException {
+    String[] line;
+    String[] types = new String[this.header.length];
+    while ((line = reader.readNext()) != null
+        && reader.getLinesRead() < this.numberOfRowsToDetermineTypeBy) {
+      int i = 0;
+      for (String value : line) {
+        // determine type
+        // if value is NA, it's not a string by definition
+        if (!value.isEmpty()) {
+          String type = getTypeOfCell(value);
+          // only set to double if is not string yet, because if one value is not double, the cell
+          // cannot be double and must be string
+          if (Array.get(types, i) != STRING && Objects.equals(type, DOUBLE)
```

Comment:
we only have double and string in parquet? no int, no categorical?

Reply:
Yes, I have looked at parquet files that have categoricals/ints/doubles in them and found out only two datatypes of the parquet files were used.